### PR TITLE
MoE Grouped MM support (5X+ MoE training perf gains)

### PR DESCRIPTION
**Pull request by lhl**
Adds support for grouped MoE kernels that replaces HF Transformers mega slow expert loop using either Torch's grouped_mm or the TRL [kernels-community/megablocks](https://huggingface.co/kernels-community/megablocks) or my [shisa-ai/megablocks/hip](https://huggingface.co/shisa-ai/megablocks-hip) port.

Support for:
- Qwen3 MoE
- Bailing V2 (Ling/Ring 2.0)
- GLM4 

Can be easily extended. I only tested the first two but they work.

  - Adds monkeypatch modules that stack expert weights, run routing with torch._grouped_mm, and fall back to MegaBlocks HIP kernels or per-expert matmuls when the grouped op isn’t available
  - Multi-GPU/DeepSpeed friendly - shared helper (clone_expert_parameter) understands ZeRO-partitioned tensors so expert weights can be stacked even under ZeRO-3
  - This should work on NVIDIA or AMD (all tests on MI300X node)
  - Config schema grows mlp_impl/use_grouped_moe_kernels toggles; docs/AMD guide describe how to build/run the MegaBlocks HIP backend on MI300X.
  - Test coverage: unit tests for Bailing grouped MLP parity and Qwen3 grouped parity; a comprehensive Ring grouped-vs-vanilla parity suite (plus optional MegaBlocks long/grad runs) to ensure loss traces stay aligned; smoke tests for Qwen3/Ring training with grouped kernels.

  Testing

  - PYTHONPATH=src pytest tests/unit/test_bailing_moe_grouped.py
  - PYTHONPATH=src pytest tests/unit/test_qwen3_moe_grouped.py
  - PYTHONPATH=src pytest tests/e2e/test_ring_moe_grouped.py -k grouped_matches_vanilla
  - Full-model validation already done with Qwen3_MoE configs (grouped vs vanilla numerics match).

Numerics on runs on my tests match exactly, just faster:

W B Chart 11_12_2025, 11_41_21 AM

## Summary by CodeRabbit

* **New Features**
  * Grouped MoE kernel support for faster training and inference on supported MoE architectures
  * New configuration options to select MoE backend implementations and enable grouped MoE kernels
  * MI300X optimization support with MegaBlocks grouped MoE

* **Documentation**
  * Added MI300X MegaBlocks grouped MoE configuration guide with validation examples

### CODE REVIEW

**Comment by lhl** [comment]
FYI busy today but I'll review the coderabbit review items one by one and see if they're actual problems or not.

File: src/axolotl/monkeypatch/models/qwen3_moe/modeling.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+            start_idx = 0
+            bias_iter = bias_bank if bias_bank is not None else [None] * weight_bank.size(0)
+            for count, weight, bias in zip(repeat_sizes.tolist(), weight_bank, bias_iter):
+                if count == 0:
+                    continue
+                end_idx = start_idx + count
+                chunk = inputs[start_idx:end_idx]
+                result = torch.matmul(chunk, weight.transpose(0, 1))
+                if bias is not None:
+                    result = result + bias
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/axolotl/monkeypatch/models/qwen3_moe/modeling.py]: OK fix pushed.

File: tests/e2e/test_ring_moe_grouped.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        from axolotl.monkeypatch.models.bailing_moe_v2.modeling import BailingMoeV2GroupedExperts
+
+        if self._orig_forward is not None:
+            BailingMoeV2GroupedExperts.forward = self._orig_forward  # type: ignore[assignment]
+
+    def summary(self) -> Optional[dict[str, Any]]:
+        if not self._per_step:
```
[the remark this answers is not in the document]
Comment (lhl) [reply tests/e2e/test_ring_moe_grouped.py:93]: added optional ForwardFn that should resolve nonetype linting error

**Comment by lhl** [comment]
OK all the linting fixed, should be gtg. I also added GLM4 by community request. Haven't run it but it passed unit and e2e tests.

**Comment by winglian** [comment]
Awesome PR @lhl ! We'll probably have to refactor this a bit after https://github.com/huggingface/transformers/pull/41580 lands upstream as well.

**Comment by winglian** [comment]
When the trainer saves checkpoints, I assume they are shaped for the patched modeling code? If so, is a conversion script needed to convert it back to the original shapes so it is usable elsewhere for inference?

**Comment by lhl** [comment]
> Awesome PR @lhl ! We'll probably have to refactor this a bit after [huggingface/transformers#41580](https://github.com/huggingface/transformers/pull/41580) lands upstream as well.

No worries, happy to help on that when it happens. I'm actively using this path for MoE training I'm invested in making sure this is actually working :)

**Comment by lhl** [comment]
> When the trainer saves checkpoints, I assume they are shaped for the patched modeling code? If so, is a conversion script needed to convert it back to the original shapes so it is usable elsewhere for inference?

So, with megablocks, the safetensor outputs for the checkpoints and final model are all good - experts are all ungrouped and since my AMD hardware I was deving on didn't have pytorch grouped_mm support, I never went down that path.

I have some nvidia GPUs now, let me get on that and make sure that it works, will need to add a reverse conversion hook for grouped vs megablocks. Actually, should test perf on that in any case... Let me do some work over the weekend.

BTW if the transformers update has landed and is going into a release, can coordinate to just do refactor and do this as an all in one to make sure we have something clean going to upstream.

**Comment by winglian** [comment]
@lhl now that https://github.com/huggingface/transformers/pull/41580 has landed, the qwen3 moe for example has a layout closer to what could be supported with megabucks. https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L215-L286
